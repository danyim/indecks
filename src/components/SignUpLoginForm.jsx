import React from 'react'
import PropTypes from 'prop-types'
import styles from '../styles/components/SignUpLoginForm.styl'

class SignUpLoginForm extends React.Component {
  static propTypes = {
    login: PropTypes.func.isRequired,
    signUpEmail: PropTypes.func.isRequired,
    signInGithub: PropTypes.func.isRequired, // eslint-disable-line
    signInTwitter: PropTypes.func.isRequired, // eslint-disable-line
    signInGoogle: PropTypes.func.isRequired, // eslint-disable-line
    isAuthenticating: PropTypes.bool,
  }

  static defaultProps = {
    isAuthenticating: false,
  }

  static providerMap = {
    GITHUB: 'signInGithub',
    TWITTER: 'signInTwitter',
    GOOGLE: 'signInGoogle',
  }

  state = {
    username: '',
    password: '',
    message: '',
  }

  constructor(props) {
    super(props)

    this.handleChange = this.handleChange.bind(this)
    this.handleLoginClick = this.handleLoginClick.bind(this)
    this.handleSignUpClick = this.handleSignUpClick.bind(this)
    this.handleClickProviderAuth = this.handleClickProviderAuth.bind(this)
  }

  handleChange(e) {
    const target = e.target
    const value = target.type === 'checkbox' ? target.checked : target.value
    const name = target.name

    this.setState({
      [name]: value,
    })
  }

  handleLoginClick(e) {
    e.preventDefault()

    if (this.state.username === '' || this.state.password === '') {
      this.setState({ message: 'Username or password cannot be blank' })
      return
    }

    this.props.login(this.state.username, this.state.password).then(message => {
      if (message) {
        this.setState({ message })
      }
    })
  }

  handleSignUpClick(e) {
    e.preventDefault()

    if (this.state.username === '' || this.state.password === '') {
      this.setState({ message: 'Username or password cannot be blank' })
      return
    }

    this.props.signUpEmail(this.state.username, this.state.password).then(message => {
      if (message) {
        this.setState({ message })
      }
    })
  }

  handleClickProviderAuth(e, provider = null) {
    e.preventDefault()

    if (!provider) {
      console.error('Provider cannot be blank')
      return
    }

    const providerSignIn = this.props[SignUpLoginForm.providerMap[provider]]

    providerSignIn()
      .then(message => {
        if (message) {
          this.setState({ message })
        }
      })
      .catch(err => {
        console.error('error while trying to log on with provider', provider, err)
      })
  }

  render() {
    return (
      <form className={styles.form}>
        <div className={styles.error}>{this.state.message}</div>
        <label htmlFor="username">
          <span>Email</span>
          <input
            name="username"
            type="text"
            placeholder="jane.doe@gmail.com"
            onChange={this.handleChange}
          />
        </label>
        <label htmlFor="password">
          <span>Password</span>
          <input
            name="password"
            type="password"
            placeholder="*********"
            onChange={this.handleChange}
          />
        </label>

        <div className={`${styles['control-buttons']}`}>
          <button
            name="login"
            type="button"
            className="button outline"
            disabled={this.props.isAuthenticating}
            onClick={this.handleLoginClick}
          >
            Login
          </button>
          <button
            name="signup"
            type="button"
            className="button"
            disabled={this.props.isAuthenticating}
            onClick={this.handleSignUpClick}
          >
            Sign up using email
          </button>
        </div>

        <div className="txt-center m-t">
          <p>or sign in using a provider below</p>
        </div>

        <div className={`${styles['control-buttons']}`}>
          <button
            type="button"
            className="button github"
            disabled={this.props.isAuthenticating}
            onClick={e => this.handleClickProviderAuth(e, 'GITHUB')}
          >
            <i className="fa fa-github fa-2x" />
            GitHub
          </button>
          <button
            type="button"
            className="button google"
            disabled={this.props.isAuthenticating}
            onClick={e => this.handleClickProviderAuth(e, 'GOOGLE')}
          >
            <i className="fa fa-google fa-2x" />
            Google
          </button>
          <button
            type="button"
            className="button twitter"
            disabled={this.props.isAuthenticating}
            onClick={e => this.handleClickProviderAuth(e, 'TWITTER')}
          >
            <i className="fa fa-twitter fa-2x" />
            Twitter
          </button>
        </div>
      </form>
    )
  }
}

export default SignUpLoginForm
