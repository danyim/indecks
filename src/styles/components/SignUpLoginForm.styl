@import '../_variables'

// figure is added here to give precedence over the base grid-figure
figure.grid-figure
  display: flex
  min-height: 275px
  flex-flow: column nowrap
  justify-content: center

.form
  input
    display: block
    margin: 1rem auto

.error
  color: red
  font-weight: bold
  text-align: center
  margin: 1rem 0

.control-buttons
  flex: 0 0 auto
  flex-flow: row nowrap
  margin-top: 1.5rem
  button
    margin: 0 0 0 1rem
    i
      display: block
      margin-bottom: 0.5rem
  button:first-child
    margin: 0
  :global
    button.twitter
      color: #00aced
      &:hover
        border-color: #00aced
    button.google
      color: #dd4b39
      &:hover
        border-color: #dd4b39
    button.github
      color: #333
      &:hover
        border-color: #333

@media (max-width: cardWidth)
  .control-buttons
    flex: 0 0 50%
    flex-flow: column nowrap
    align-content: space-around
    button
      margin: .6rem 0 0
    button:first-child
      margin: 0
