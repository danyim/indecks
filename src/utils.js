export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

