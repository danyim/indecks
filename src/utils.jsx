import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
