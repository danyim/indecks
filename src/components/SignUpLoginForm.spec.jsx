import toJson from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import SignUpLoginForm from './SignUpLoginForm'

describe('SignUpLoginForm', () => {
  let props

  beforeEach(() => {
    props = {
      login: () => {},
      signUpEmail: () => {},
      signInGithub: () => {},
      signInTwitter: () => {},
      signInGoogle: () => {},
      isAuthenticating: false,
    }
  })

  it('should render self and subcomponents', () => {
    const wrapper = shallow(SignUpLoginForm, props)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should display the form elements', () => {
    const { wrapper } = setup(SignUpLoginForm, props)

    const inputEmail = wrapper.find('input[name="username"]')
    const inputPass = wrapper.find('input[name="password"]')

    expect(inputEmail.exists()).toBe(true)
    expect(inputPass.exists()).toBe(true)
  })

  it('should call login if Login is clicked', () => {
    const username = 'testEmail@test.com'
    const password = 'test_pass'

    const handler = jest.fn().mockReturnValue({ then: () => {} })
    const { wrapper } = setup(SignUpLoginForm, {
      ...props,
      login: handler,
    })

    const inputEmail = wrapper.find('input[name="username"]')
    const inputPass = wrapper.find('input[name="password"]')
    const buttonLogin = wrapper.find('button[name="login"]')

    inputEmail.simulate('change', {
      target: { value: username, name: 'username' },
    })
    inputPass.simulate('change', {
      target: { value: password, name: 'password' },
    })
    buttonLogin.simulate('click', { preventDefault: () => {} })
    expect(handler).toHaveBeenCalled()
  })

  it('should call signUpEmail if signup is clicked', () => {
    const username = 'testEmail@test.com'
    const password = 'test_pass'

    const handler = jest.fn().mockReturnValue({ then: () => {} })
    const { wrapper } = setup(SignUpLoginForm, {
      ...props,
      signUpEmail: handler,
    })

    const inputEmail = wrapper.find('input[name="username"]')
    const inputPass = wrapper.find('input[name="password"]')
    const buttonSignup = wrapper.find('button[name="signup"]')

    inputEmail.simulate('change', {
      target: { value: username, name: 'username' },
    })
    inputPass.simulate('change', {
      target: { value: password, name: 'password' },
    })
    buttonSignup.simulate('click', { preventDefault: () => {} })
    expect(handler).toHaveBeenCalled()
  })

  it('should call signInGithub if the Github provider button is clicked', () => {
    const handler = jest
      .fn()
      .mockReturnValue({ then: () => ({ catch: () => {} }) })
    const { wrapper } = setup(SignUpLoginForm, {
      ...props,
      signInGithub: handler,
    })

    const button = wrapper.find('button.github')

    button.simulate('click', { preventDefault: () => {} })
    expect(handler).toHaveBeenCalled()
  })

  it('should call signInTwitter if the Twitter provider button is clicked', () => {
    const handler = jest
      .fn()
      .mockReturnValue({ then: () => ({ catch: () => {} }) })
    const { wrapper } = setup(SignUpLoginForm, {
      ...props,
      signInTwitter: handler,
    })

    const button = wrapper.find('button.twitter')

    button.simulate('click', { preventDefault: () => {} })
    expect(handler).toHaveBeenCalled()
  })

  it('should call signInGoogle if the Google provider button is clicked', () => {
    const handler = jest
      .fn()
      .mockReturnValue({ then: () => ({ catch: () => {} }) })
    const { wrapper } = setup(SignUpLoginForm, {
      ...props,
      signInGoogle: handler,
    })

    const button = wrapper.find('button.google')

    button.simulate('click', { preventDefault: () => {} })
    expect(handler).toHaveBeenCalled()
  })

  describe('Validation', () => {
    let handler
    const username = 'testEmail@test.com'
    const password = 'test_pass'
    const testButtons = [
      {
        title: 'logging in',
        selector: 'button[name="signup"]',
      },
      {
        title: 'signing up',
        selector: 'button[name="signup"]',
      },
    ]

    beforeEach(() => {
      handler = jest.fn().mockReturnValue({ then: () => {} })
    })

    testButtons.forEach(testButton => {
      it(`should not allow empty usernames for the username and password when ${
        testButton.title
      }`, () => {
        const { wrapper } = setup(SignUpLoginForm, {
          ...props,
          signUpEmail: handler,
        })

        const button = wrapper.find(testButton.selector)

        // Empty form
        button.simulate('click', { preventDefault: () => {} })
        expect(handler).not.toHaveBeenCalled()
        expect(wrapper.state('message') !== '').toBe(true)
      })

      it(`should not allow empty passwords when ${testButton.title}`, () => {
        const { wrapper } = setup(SignUpLoginForm, {
          ...props,
          signUpEmail: handler,
        })

        const inputEmail = wrapper.find('input[name="username"]')
        const button = wrapper.find(testButton.selector)

        inputEmail.simulate('change', {
          target: { value: username, name: 'username' },
        })
        button.simulate('click', { preventDefault: () => {} })
        expect(handler).not.toHaveBeenCalled()
        expect(wrapper.state('message') !== '').toBe(true)
      })

      it(`should not allow empty usernames when ${testButton.title}`, () => {
        const { wrapper } = setup(SignUpLoginForm, {
          ...props,
          signUpEmail: handler,
        })

        const inputPass = wrapper.find('input[name="password"]')
        const button = wrapper.find(testButton.selector)

        inputPass.simulate('change', {
          target: { value: password, name: 'password' },
        })
        button.simulate('click', { preventDefault: () => {} })
        expect(handler).not.toHaveBeenCalled()
        expect(wrapper.state('message') !== '').toBe(true)
      })
    })
  })
})
