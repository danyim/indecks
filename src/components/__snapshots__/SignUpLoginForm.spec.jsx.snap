// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpLoginForm should render self and subcomponents 1`] = `
<form
  className="form"
>
  <div
    className="error"
  />
  <label
    htmlFor="username"
  >
    <span>
      Email
    </span>
    <input
      name="username"
      onChange={[Function]}
      placeholder="jane.doe@gmail.com"
      type="text"
    />
  </label>
  <label
    htmlFor="password"
  >
    <span>
      Password
    </span>
    <input
      name="password"
      onChange={[Function]}
      placeholder="*********"
      type="password"
    />
  </label>
  <div
    className="control-buttons"
  >
    <button
      className="button outline"
      disabled={false}
      name="login"
      onClick={[Function]}
      type="button"
    >
      Login
    </button>
    <button
      className="button"
      disabled={false}
      name="signup"
      onClick={[Function]}
      type="button"
    >
      Sign up using email
    </button>
  </div>
  <div
    className="txt-center m-t"
  >
    <p>
      or sign in using a provider below
    </p>
  </div>
  <div
    className="control-buttons"
  >
    <button
      className="button github"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-github fa-2x"
      />
      GitHub
    </button>
    <button
      className="button google"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-google fa-2x"
      />
      Google
    </button>
    <button
      className="button twitter"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-twitter fa-2x"
      />
      Twitter
    </button>
  </div>
</form>
`
